### PR TITLE
Issue 225 error checking

### PR DESCRIPTION
**Pull request by JeffreyZh4ng**
https://github.com/ILLIXR/ILLIXR/issues/225

### CODE REVIEW

File: configs/core_plugins.yaml
Comment (e3m3) [inline configs/core_plugins.yaml:6]:
```suggestion
      version: eb0b3ded1918567b4ff5bac1078e6bdd4a8baee1
```

File: configs/rt_slam_plugins.yaml
Comment (e3m3) [inline configs/rt_slam_plugins.yaml]:
```suggestion
      version: 648ac892543701abd78a6f4bc56a6b6001f8413f
```

**Review by e3m3 — CHANGES_REQUESTED** [review]
Updated the commit numbers to reflect updated branches for Kimera-VIO and audio_pipeline

File: configs/rt_slam_plugins.yaml
Comment (charmoniumQ) [reply configs/rt_slam_plugins.yaml]: I suggest we use a better strategy for managing these commit references.

ILLIXR/ILLIXR (what I'll call "main repo") needs to refer to ILLIXR/audio_pipeline and friends (what I'll call "subrepos"). Currently, we do this by pinning an exact version (commit hash). If the subrepo updates, the main repo does not automatically get that change. This is good for long-term stability, bad for rapid development. If you do want those changes, one has to do these small edits. However, these tiny changes are easily overlooked (as they were here) and can be easily wrong.

Referring to "master" of the subrepos is on the other end of the spectrum. It's great for rapid development and bad for stability. We know that branches of main repo can fall months out-of-date. If the main repo references master of the subrepo, the subrepos will have to support a wide range of main repo versions. If a subrepo updates master, it could break old branches of the main repo, which get the update automatically.

The in-between solution I propose is managing these by semver. Whenever master in a subrepo updates, its maintainer may choose to move the tag x.y to point to the new commit or create a new tag, x.(y+1) or (x+1).0. This way, we explicitly say when we are breaking backwards compatibility (when the minor or major version changes), and non-breaking changes can be pulled in for free, enabling rapid development without compromising deterministic behavior of old branches.

Comment (charmoniumQ) [reply configs/rt_slam_plugins.yaml]: image

Comment (e3m3) [reply configs/rt_slam_plugins.yaml]: I agree that we should move to a tag-based system like you suggest. @mhuzai Should the versioning system we use match the milestone release versions to avoid confusion? In that case, we would start from a version `2.X` tag?

Comment (mhuzai) [reply configs/rt_slam_plugins.yaml]: Well, the components are independent of each other, so they should be able to bump their versions independently. We can tag the last standalone commit 1.0, the current latest commit 2.0, and then bump versions in each subrepo as required.

**Comment by mhuzai** [comment]
Please remove duplicate commits from other branches.

**Review by e3m3 — APPROVED** [review]

**Review by charmoniumQ — APPROVED** [review]

File: timewarp_gl/plugin.cpp
Comment (charmoniumQ) [inline timewarp_gl/plugin.cpp:678]: `free` is good, but `std::vector` is better for these variables.

I think we want them to be a vector of vectors, one for each eye. This eliminates `vec[eye*NUM_ELEM + elem]`.

File: runtime/sqlite_record_logger.hpp
Comment (charmoniumQ) [inline runtime/sqlite_record_logger.hpp:133]: Yup. Good call.

File: offline_imu_cam/plugin.cpp
Comment (charmoniumQ) [inline offline_imu_cam/plugin.cpp:101]: IMO unnecessary as `offline_imu_cam` is tightly-coupled to `lazy_load` already. It can be part of the "contract" that lazy loud emits grayscale images. I would rather just `assert` that this contract is met in `offline_imu_cam`.

```
assert(blah->channels() == 1 && "Got a color image where grayscale is expected");
```

File: gldemo/plugin.cpp
Comment (charmoniumQ) [inline gldemo/plugin.cpp:403]: At faults, I think `abort` is better because it fires a `SIGABRT` signal, which can be caught by a debugger. `std::exit` and `std::terminate` cannot, AFAIK. I suppose this is why `assert` sends a `SIGABRT`.

The question is, should destructors be run if the exit is due to a fatal error? I think the answer is no. In the case of a fatal error, the output is likely "tainted" by the error. In the _special case_ that the operator does want it, they can replace that particular `abort` with an `std::exit`, but I contend this should not be the default,

File: gldemo/plugin.cpp
Comment (charmoniumQ) [inline gldemo/plugin.cpp:354]: I think it's best to `assert(0 && "something bad happened")`. It's less typing anyway.

Unfortunately, `assert(glXstuff() ...)` (eliding the if-statement) won't work because it won't run the expression when `profile = opt`.

File: gldemo/plugin.cpp
Comment (charmoniumQ) [inline gldemo/plugin.cpp:68]: :+1:

File: debugview/plugin.cpp
Comment (charmoniumQ) [inline debugview/plugin.cpp:57]: Shall we exit/abort in this case?

File: common/shader_util.hpp
Comment (charmoniumQ) [inline common/shader_util.hpp:31]: Shall we exit/abort in this case?

File: common/extended_window.hpp
Comment (charmoniumQ) [inline common/extended_window.hpp:43]: The point of this is that non-zero `errno` here doesn't actually denote an erroneous condition.

Any output caused by this is noise. Let's go back to `errno = 0`.

File: offline_imu_cam/plugin.cpp
Comment (e3m3) [reply offline_imu_cam/plugin.cpp:101]: In previous testing, calling `cv:cvtColor` was an issue because the input was not color. In testing this branch, I had the opposite issue- removing the call was an issue because the input was indeed in color. Did the behavior of `lazy_load` change recently?

File: gldemo/plugin.cpp
Comment (e3m3) [reply gldemo/plugin.cpp:403]: `abort` does not allow the C++ runtime to perform cleanup:
>  The difference between exit and abort is that exit allows the C++ run-time termination processing to take place (global object destructors get called), but abort terminates the program immediately. The abort function bypasses the normal destruction process for initialized global static objects. It also bypasses any special processing that was specified using the atexit function.
[Source](https://docs.microsoft.com/en-us/cpp/cpp/program-termination?view=msvc-160). It seems to me that `abort` could be useful for a debugging environment, though. Perhaps we can use a small utility that switches the `std::exit` to `abort` if `#ifdndef NDEBUG`?

Comment (e3m3) [reply gldemo/plugin.cpp:403]: In terms of the destructors, we can be holding onto stateful resources (e.g., file handles) which should be closed normally if possible to avoid odd side-effects.

Comment (e3m3) [reply gldemo/plugin.cpp:403]: That is, I think debugging via `SIGABRT` is the special case, given that the user will need to resolve the failure.

File: gldemo/plugin.cpp
Comment (e3m3) [reply gldemo/plugin.cpp:354]: Agreed. The value can be stored temporarily.

File: common/extended_window.hpp
Comment (e3m3) [reply common/extended_window.hpp:43]: Agreed the output is noise, but currently, `RAC_ERRNO` and `RAC_ERRNO_MSG` only print and clear when `NDEBUG` is on. Errnos throughout execution during a run with `opt` are ignored. Clearing errno here would be the exception.

File: common/global_module_defs.hpp
Comment (e3m3) [reply common/global_module_defs.hpp:30]: The meaning for `inline` has somewhat changed to mean that multiple definitions are allowed, but `static` probably results in the same thing but with a cleaner output.

Comment (e3m3) [reply common/global_module_defs.hpp:30]: The meaning for `inline` has somewhat changed to mean that multiple definitions are allowed, but `static` probably results in the same thing but with a cleaner output.

File: common/extended_window.hpp
Comment (charmoniumQ) [reply common/extended_window.hpp:43]: Debug mode will be active when someone is hunting for the source of a true bug. Getting a print for a not-genuine-error in such case would be misleading. The reader has to read the message and understand the source to determine that it is not genuine.

File: offline_imu_cam/plugin.cpp
Comment (charmoniumQ) [reply offline_imu_cam/plugin.cpp:101]: No, I don't think it changed recently.

Since the component is only used here, the behavior "should" be whatever makes it easier for `offline_imu_cam` and its downstream, `open_vins`. I believe loading grayscale images from `cv::imread` is the right way to go.

File: common/common.mk
Comment (e3m3) [reply common/common.mk:6]: Yep, I think `catchsegv` uses `backtrace`.

**Review by charmoniumQ — DISMISSED** [review]

File: gldemo/plugin.cpp
Comment (charmoniumQ) [inline gldemo/plugin.cpp:358]: std::cerr << "GLEW error: " << glewGetErrorString(err) << std::endl;

File: common/shader_util.hpp
Comment (charmoniumQ) [inline common/shader_util.hpp:111]: I believe these `std::exit`s should be `illixr::abort`s.

File: offline_imu_cam/data_loading.hpp
Comment (charmoniumQ) [inline offline_imu_cam/data_loading.hpp:27]: Remove `TODO`

**Comment by charmoniumQ** [comment]
Also note: #44 relates to using a "logging infrastructure for human-readable messages" such as [glog](https://github.com/google/glog). Glog has `PLOG` which prints your message, the line number, and the human-readable message associated with errno.

I vote to accept this now, but have an eye towards upgrading these facilities in #44. We could upgrade the implementation of your error-reporting macros without changing their call-sites.

File: offline_imu_cam/data_loading.hpp
Comment (charmoniumQ) [inline offline_imu_cam/data_loading.hpp:29]: One more request :laughing: :
Delete this comment and the commented code; it is now out-of-date.

Comment (e3m3) [reply offline_imu_cam/data_loading.hpp:29]: You got it

**Review by charmoniumQ — DISMISSED** [review]

File: timewarp_gl/plugin.cpp
Comment (mhuzai) [inline timewarp_gl/plugin.cpp:506]: Only do this when collecting data / offload is enabled.

File: timewarp_gl/plugin.cpp
Comment (mhuzai) [inline timewarp_gl/plugin.cpp:527]: Why is this necessary? Won't GL errors get reported via the debug callback?

File: timewarp_gl/plugin.cpp
Comment (mhuzai) [inline timewarp_gl/plugin.cpp:673]: Why is this necessary, given line 665?

File: timewarp_gl/plugin.cpp
Comment (mhuzai) [inline timewarp_gl/plugin.cpp:738]: This wording has always confused me. Can you make it something like "Next swap in: xyz ms"?

File: timewarp_gl/plugin.cpp
Comment (mhuzai) [inline timewarp_gl/plugin.cpp:777]: Don't we take care of everything at this point? If so, remove this comment.

**Review by mhuzai — COMMENTED** [review]
I am still not sure what the exact combo of asserts and RACs should be for errno checking. Let's discuss that via chat.

File: common/extended_window.hpp
Comment (e3m3) [reply common/extended_window.hpp:36]: Its better to be explicit, and avoid relying on `using namespace ...` elsewhere in the code. If the `using` line gets deleted or moved, the code breaks. Also, even with `using`, the code does not need any context for the reader to understand that it's not a call to `std::abort`. Also see [SO](https://stackoverflow.com/questions/1452721/why-is-using-namespace-std-considered-bad-practice) and this [article](https://www.geeksforgeeks.org/using-namespace-std-considered-bad-practice/).

File: common/extended_window.hpp
Comment (e3m3) [reply common/extended_window.hpp:32]: Agreed. We will tackle this in #44.

> Also note: #44 relates to using a "logging infrastructure for human-readable messages" such as glog. Glog has PLOG which prints your message, the line number, and the human-readable message associated with errno.

> I vote to accept this now, but have an eye towards upgrading these facilities in #44. We could upgrade the implementation of your error-reporting macros without changing their call-sites.

File: configs/rt_slam_plugins.yaml
Comment (e3m3) [reply configs/rt_slam_plugins.yaml:7]: It hasn't but this is an update to be merged that exists in the Kimera-VIO repo now to accompany this PR. Since the tag for the corresponding commit exists, we can test this PR in CI, and then merge the equivalent Kimera-VIO PR somewhat independently.

All versions will be easier to track once we integrate the refactored configurations for #231 and #54. Each external module has its own `.yaml` configuration specifying the tag so we can update versions in a consistent manner across plugins and configurations.

File: runtime/sqlite_record_logger.hpp
Comment (e3m3) [reply runtime/sqlite_record_logger.hpp:51]: It's just not clear. Not everyone is madly in love with C/C++ like we are :smiley:

File: runner/runner/main.py
Comment (e3m3) [reply runner/runner/main.py:105]: This gets reverted in #198.

File: timewarp_gl/plugin.cpp
Comment (e3m3) [reply timewarp_gl/plugin.cpp:527]: We can remove it. This one was just for testing/abundance of caution.

File: common/shader_util.hpp
Comment (e3m3) [reply common/shader_util.hpp:50]: I removed the numbers, but left the print so that we can get info from `msg`. Can remove also.

File: common/shader_util.hpp
Comment (e3m3) [reply common/shader_util.hpp:32]: We check more explicitly using the `severity` interface above. Continuing to use `severity` over `type` might be more robust.

Either we can:
- Return early for `GL_DEBUG_SEVERITY_LOW` and `GL_DEBUG_SEVERITY_MEDIUM` like for `GL_DEBUG_SEVERITY_NOTIFICATION`, or
- Only return early for `GL_DEBUG_SEVERITY_NOTIFICATION`, and guard the `abort` for `GL_DEBUG_SEVERITY_LOW` and `GL_DEBUG_SEVERITY_MEDIUM` so that we get the print out for low, medium, and high, but only crash when it's high.

File: timewarp_gl/plugin.cpp
Comment (e3m3) [reply timewarp_gl/plugin.cpp:5]: Even if it was, we should try to include everything used within a compilation unit (file) so that we can easily implement #213 and #54.

**Review by charmoniumQ — DISMISSED** [review]

File: common/shader_util.hpp
Comment (mhuzai) [reply common/shader_util.hpp:33]: Option 2 sounds great, but let's crash with medium too. This is just to err on the side of caution.

File: configs/rt_slam_plugins.yaml
Comment (mhuzai) [reply configs/rt_slam_plugins.yaml:7]: Nice!

**Review by charmoniumQ — DISMISSED** [review]

**Review by mhuzai — APPROVED** [review]

File: runner.sh
Comment (charmoniumQ) [inline runner.sh:10]: Let's put this in the `build_one_plugin`. If someone wants to know how to "build one plugin," they will look there, so it is more discoverable there.

**Review by charmoniumQ — COMMENTED** [review]
Also, how many "extra" stdout messages does this generate?

**Comment by e3m3** [comment]
> Also, how many "extra" stdout messages does this generate?

Every page or so of the `dbg`-enabled output will have about the following amount of text:
```
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
|> Message: switchboard_impl after wait_dequeue_timed
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
|> Message: switchboard_impl after wait_dequeue_timed
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
|> Message: switchboard_impl after wait_dequeue_timed
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
|> Message: switchboard_impl after wait_dequeue_timed
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
|> Message: switchboard_impl after wait_dequeue_timed
thread,140643395774208,opencv threadpool worker,0
|| Errno was set: 110 @ ./switchboard_impl.hpp:307[check_queues]
```
I'm not too concerned about the amount of output here, but it will be dependent on how many future `errno` setting calls we make. I can add a quieting flag here if we want.

**Review by charmoniumQ — APPROVED** [review]
Approval, conditioned on it passing CI some day.

**Review by charmoniumQ — APPROVED** [review]

**Review by mhuzai — APPROVED** [review]

**Pull request merged**
